src/Evaluator.hs

```diff
@@ -38,8 +103,343 @@ data EvalError
   | EvalErrorType TypeError
   | EvalErrorCase SrcLoc Value
   | EvalErrorUser Syntax
-  deriving (Show)
+  | EvalErrorIdent Value
+  deriving (Show, Typeable)
 makePrisms ''EvalError
+instance Exception EvalError
+
+-- | The Kontinuation type. The naming convention InFoo means that the subject
+-- of evaluation in the CEK machine is Foo. For example, when the continuation
+-- is 'InArg' the subject of evaluation is the argument of the function and the
+-- continuation holds the un-evaluated function symbol in its first field.
+data Kont where
+
+  Halt :: Kont
+  -- ^ Marks the evaluator finishing
+
+  -- functions
+  InArg :: !(CoreF TypePattern ConstructorPattern Core) -> !VEnv -> !Kont -> Kont
+  -- ^ The argument is being evaluated, so hold onto the function symbol.
+  InFun :: !Value -> !VEnv -> !Kont -> Kont
+  -- ^ The function is being evaluated, so hold onto the evaluated argument.
```

Comment:
These comments are really helpful, let's add one for each constructor!

```
  -- functions
  InArg :: !(CoreF TypePattern ConstructorPattern Core) -> !VEnv -> !Kont -> Kont
  -- ^ The argument is being evaluated, so hold onto the function expression
  InFun :: !Value -> !VEnv -> !Kont -> Kont
  -- ^ The function is being evaluated, so hold onto the evaluated argument.

  InLetDef :: !Ident -> !Var -> !(CoreF TypePattern ConstructorPattern Core) -> !VEnv -> !Kont -> Kont
  -- ^ Inside the rhs of (let [x rhs] body), so hold the name x and the body.

  -- constructors
  InCtor :: !Constructor -> ![CoreF TypePattern ConstructorPattern Core] -> ![Value] -> !VEnv -> !Kont -> Kont
  -- ^ Inside arg_i of (Ctor arg_1 ... arg_i ... arg_n), so hold Ctor, the
  -- already evaluated values value_1 ... value_{i-1}, and the
  -- yet-to-be-evaluated expressions arg_{i+1} ... arg_n.

  -- Cases
  InCaseScrut     :: ![(SyntaxPattern, Core)] -> !SrcLoc -> !VEnv -> !Kont -> Kont
  InDataCaseScrut :: ![(ConstructorPattern, Core)] -> !SrcLoc -> !VEnv -> !Kont -> Kont
  InTypeCaseScrut :: ![(TypePattern, Core)] -> !SrcLoc -> !VEnv -> !Kont -> Kont
  -- ^ Inside the scrutinee of (case scrutinee [pat_1 rhs_1] ...), so hold
  -- the branches.

  -- lists
  InConsHd :: !Core -> !(CoreF TypePattern ConstructorPattern Core) -> !VEnv -> !Kont -> Kont
  -- ^ Inside the head of (cons-list-syntax-cons head tail loc), so hold the tail.
  InConsTl :: !Core -> !Syntax -> !VEnv -> !Kont -> Kont
  -- ^ Inside the tail of (cons-list-syntax-cons head tail loc), so hold the already-evaluated head.
  InList   :: !Core -> ![Core] -> ![Syntax] -> !VEnv -> !Kont -> Kont
  -- ^ Inside arg_i of (list-syntax (arg_1 ... arg_i ... arg_n) loc), so hold
  -- the already-evaluated syntax values stx_1 ... stx_{i-1}, and the
  -- yet-to-be-evaluated expressions arg_{i+1} ... arg_n.

  -- idents
  InIdent :: !Core -> !VEnv -> !Kont -> Kont
  InIdentEqL :: !HowEq -> !Core -> !VEnv -> !Kont -> Kont
  -- ^ Inside the lhs of (<how>-identifier=? lhs rhs), so hold <how> and rhs.
  InIdentEqR :: !HowEq -> !Value -> !VEnv -> !Kont -> Kont
  -- ^ Inside the rhs of (<how>-identifier=? lhs rhs), so hold <how> and the
  -- already-evaluated lhs.

  -- Macros
  InPureMacro :: !VEnv -> !Kont -> Kont
  -- ^ Inside the expr of (pure expr), so nothing to hold.
  InBindMacroHd :: !Core -> !VEnv -> !Kont -> Kont
  -- ^ Inside the expr1 of (>>= expr1 expr2), so hold expr2.
  InBindMacroTl :: !MacroAction -> !VEnv -> !Kont -> Kont
  -- ^ Inside the expr2 of (>>= expr1 expr2), so hold the already-evaluated
  -- expr1.

  -- atomics
  InInteger :: !Core -> !VEnv -> !Kont -> Kont
  InString  :: !Core -> !VEnv -> !Kont -> Kont
  InLoc     :: !Core -> !VEnv -> !Kont -> Kont
  -- ^ Inside the loc of (replace-loc loc stx), so hold the stx.
  InLocStx  :: !SrcLoc -> !VEnv -> !Kont -> Kont
  -- ^ Inside the stx of (replace-loc loc stx), so hold the already-evaluated
  -- loc.

  -- scope
  InScope :: !(ExprF Syntax) -> !VEnv -> !Kont -> Kont
  -- ^ Inside the loc of something like (cons-syntax-list head tail loc), so
  -- remember the Syntax-Contents which needs a scope-set and location.

  -- logs and errors
  InLog   :: !VEnv -> !Kont -> Kont
  -- ^ Inside the expr of (log expr), so nothing to hold.
  InError :: !VEnv -> !Kont -> Kont
  -- ^ Inside the expr of (error expr), so nothing to hold.
```

Reply:
turns out I was wrong about some of those, as some of those hold a reversed list.

:pencil2: this is not a blocker, but after this MR is finally merged, I plan to come back and add those comments.

Reply:
- [ ] add comments for each constructor after merge.